src/zmq/zmqpublishnotifier.cpp

```diff
@@ -86,6 +86,14 @@ bool CZMQAbstractPublishNotifier::Initialize(void *pcontext)
             return false;
         }
 
+        const int so_keepalive_option {1};
+        rc = zmq_setsockopt(psocket, ZMQ_TCP_KEEPALIVE, &so_keepalive_option, sizeof(so_keepalive_option));
+        if (rc != 0) {
+            zmqError("Failed to set SO_KEEPALIVE");
+            zmq_close(psocket);
```

Comment:
Maybe if it fails to set SO_KEEPALIVE, it should still continue with a warning?

Reply:
When does this fail?
If `SO_KEEPALIVE` is important, I don't think ignoring errors is a good idea. Not everyone continuously monitors the logs, and it might suddenly result in a return of issue #14687.

Reply:
FWIW,
* The kernel only fails SO_KEEPALIVE setsockopt(2) if there is something assert()-level wrong with the socket (invalid file descriptor; fd is closed; fd is not a socket; basic stuff...)
* zmq_setsockopt() is not a direct wrapper, but close:  the value is stored, and setsockopt(2) is called later in zmq's src/tcp.cpp.
* zmq_setsockopt() will only return an error if there's something assert()-level wrong with the zmq structure.

Ergo, errors are serious but will not occur due to Linux network stack conditions.   Errors only occur due to programmatic error (zmq passed file fd to socket syscall) or memory corruption.

Reply:
zmq will return 0 (success) on systems without SO_KEEPALIVE